django/db/backends/mysql/introspection.py

```diff
@@ -53,6 +54,14 @@ class DatabaseIntrospection(BaseDatabaseIntrospection):
         FIELD_TYPE.VAR_STRING: "CharField",
     }
 
+    @cached_property
+    def on_delete_types(self):
+        on_delete_types = super().on_delete_types
+        if self.connection.mysql_is_mariadb:
+            # On MariaDB "NO ACTION" and "RESTRICT" are synonyms.
+            on_delete_types["RESTRICT"] = DO_NOTHING
```

Comment:
https://mariadb.com/docs/server/ha-and-performance/optimization-and-tuning/optimization-and-indexes/foreign-keys#constraints

Reply:
> I suspect it's the same on MySQL as neither support deferrable foreign key constraints which means that both RETRICT and DO_NOTHING are equivalent.

A subtle difference is mentioned between InnoDB and NDB:

> [NDB](https://dev.mysql.com/doc/refman/8.0/en/mysql-cluster.html) supports deferred checks, and NO ACTION specifies a deferred check;

Reply:
Thanks for that, TIL about NDB.

I don't think we support the NDB storage engine though, at least we don't test against it and [we explicitly mark the MySQL backend as not supporting deferred constraint checks](https://github.com/search?q=repo%3Adjango%2Fdjango%20can_defer_constraint_checks&type=code).

Reply:
We can remove `on_delete_types` in MySQL backend, but this requires minor adjustments in introspection tests.

Updated.